src/ripple/app/tx/impl/Transactor.cpp

```diff
@@ -1894,6 +1900,214 @@ Transactor::operator()()
         applied = isTecClaim(result);
     }
 
+    if (applied && view().rules().enabled(featureServiceFee) &&
+        ctx_.tx.isFieldPresent(sfServiceFee))
+        do
+        {
+            // Service fee is processed on a best-effort basis without affecting
+            // tx application. The reason is that the client completely controls
+            // the service fee that it submits with the user's txn, and
+            // therefore is already completely aware of the user's capacity to
+            // pay the fee and therefore enforcement logic is unnecessary
+            // chain-side.
+
+            STObject const& obj = const_cast<ripple::STTx&>(ctx_.tx)
+                                      .getField(sfServiceFee)
+                                      .downcast<STObject>();
+
+            // This should be enforced by template but doesn't hurt to
+            // defensively check it here.
+            if (!obj.isFieldPresent(sfDestination) ||
+                !obj.isFieldPresent(sfAmount) || obj.getCount() != 2)
+            {
+                JLOG(j_.warn())
+                    << "service fee not applied - malformed inner object.";
+                break;
+            }
+
+            auto const src = ctx_.tx.getAccountID(sfAccount);
+            auto const dst = obj.getAccountID(sfDestination);
+
+            auto const amt = obj.getFieldAmount(sfAmount);
+
+            // sanity check fields
+            if (src == dst)
+            {
+                JLOG(j_.info())
+                    << "skipping self service-fee on " << src << ".";
+                break;
+            }
+
+            if (amt <= beast::zero)
+            {
+                JLOG(j_.info())
+                    << "skipping non-positive service-fee from " << src << ".";
+                break;
+            }
```

Comment:
The only reason why ServiceFee was not used in the transaction should be due to information in the on-ledger data, such as insufficient balance, noAuth, etc. 
Those should be a tem error.